### PR TITLE
Tandem Cross-DSDF Features: Per-Node Inter-Foil Geometry for p_tan

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Gap/Stagger Spatial Bias (GSB, PR #2130) proved that explicit tandem geometry scalars help slice routing: appending (gap, stagger) as global scalars to the spatial bias MLP delivered -3.0% p_tan. However, gap/stagger describe the **global** tandem configuration — they don't encode the local relationship of each mesh node to each individual foil.

**The idea:** Compute two explicit inter-foil geometry features from existing DSDF channels and inject them as additional per-node features before standardization:

1. **`dsdf_dist_ratio`** = `dsdf2_min / (dsdf1_min + dsdf2_min + 1e-4)` — a per-node scalar in [0,1] encoding "how much of this node's neighborhood is dominated by foil-2 vs foil-1". Near foil-2: →1. Near foil-1: →0. Zero for non-tandem samples.

2. **`dsdf_rel_angle`** = `atan2(dsdf2_gy, dsdf2_gx) - atan2(dsdf1_gy, dsdf1_gx)` — the angular difference of DSDF gradient directions between foil-2 and foil-1. Encodes local foil orientation alignment. Zero for non-tandem samples.

These are **physics-motivated** features: `dist_ratio` directly correlates with inter-foil pressure channel behavior; `rel_angle` encodes geometric alignment, a key predictor of interference lift. Both are computable from existing inputs — no new data, no new parameters beyond the input projection weight for 2 extra dims.

**Prior context:** GSB (global geometry → routing) + this PR (local geometry → per-node features) are orthogonal. GSB affects how slices are allocated; cross-DSDF features affect what the model sees at each node. No duplication with alphonse #2157 (which adds global cosine similarity to the routing MLP).

**Expected impact:** -2 to -4% p_tan (similar reasoning to GSB: explicit geometry features remove implicit learning burden).

### Step 1: Compute the features (add after `_raw_gap_stagger` extraction, before standardization)

In `train.py`, locate the augmentation/preprocessing block where `_raw_gap_stagger` is extracted (around line 1668, just before `x = (x - stats["x_mean"]) / stats["x_std"]`). Add:

```python
# Tandem Cross-DSDF Features — per-node inter-foil geometry (zero for non-tandem)
if cfg.tandem_cross_dsdf:
    # Min absolute DSDF magnitude for each foil: channels 2:6 = foil-1, 6:10 = foil-2
    _dsdf1_min = x[:, :, 2:6].abs().min(dim=-1, keepdim=True).values   # [B, N, 1]
    _dsdf2_min = x[:, :, 6:10].abs().min(dim=-1, keepdim=True).values  # [B, N, 1]
    # Relative proximity: how much of each node's neighborhood is dominated by foil-2
    _dist_ratio = _dsdf2_min / (_dsdf1_min + _dsdf2_min + 1e-4)        # [B, N, 1], range [0,1]
    # Gradient direction difference: foil-2 orientation minus foil-1 orientation
    _dsdf1_angle = torch.atan2(x[:, :, 3:4], x[:, :, 2:3])  # foil-1 DSDF gradient angle
    _dsdf2_angle = torch.atan2(x[:, :, 7:8], x[:, :, 6:7])  # foil-2 DSDF gradient angle
    _rel_angle   = _dsdf2_angle - _dsdf1_angle                # [B, N, 1], range ~[-pi, pi]
    # Zero out for non-tandem samples
    _tandem_mask_f = _is_tandem.float().view(-1, 1, 1)  # [B, 1, 1]
    _dist_ratio = _dist_ratio * _tandem_mask_f
    _rel_angle  = _rel_angle  * _tandem_mask_f
    # Append to x BEFORE standardization — shapes become [B, N, 26]
    x = torch.cat([x, _dist_ratio, _rel_angle], dim=-1)
```

**⚠️ Important:** The DSDF indices assume the standard feature layout: `x[:,:,0:2]` = xy coords, `x[:,:,2:6]` = DSDF foil-1 (signed dist + gradients), `x[:,:,6:10]` = DSDF foil-2, etc. Verify this matches the actual `feature_names` order in `split_manifest.json` before running.

### Step 2: Handle the extra input dims for standardization

The `stats["x_mean"]` and `stats["x_std"]` tensors don't have entries for the 2 new channels. The cleanest fix — append zeros to mean and ones to std so the new features pass through unstandardized:

```python
if cfg.tandem_cross_dsdf:
    _extra_mean = torch.zeros(1, 1, 2, device=x.device, dtype=x.dtype)
    _extra_std  = torch.ones(1, 1, 2, device=x.device, dtype=x.dtype)
    x_mean = torch.cat([stats["x_mean"], _extra_mean], dim=-1)
    x_std  = torch.cat([stats["x_std"],  _extra_std],  dim=-1)
    x = (x - x_mean) / x_std
else:
    x = (x - stats["x_mean"]) / stats["x_std"]
```

### Step 3: Handle the model input dim

The Transolver model receives `x` of shape `[B, N, X_DIM]`. With the 2 new channels, `X_DIM` becomes 26 instead of 24. The model instantiation must use the updated dim.

Check how the model is instantiated — if it uses `x.shape[-1]` from the first batch (dynamic), it may work automatically. If it uses a static `X_DIM` config value, add:

```python
cfg.x_dim = 24 + (2 if cfg.tandem_cross_dsdf else 0)
```

and pass this to the model constructor. For `torch.compile`, the input shape must be static — ensure the model is compiled AFTER `x_dim` is determined.

### Step 4: Add config flag

In the `Config` dataclass:
```python
tandem_cross_dsdf: bool = False
```

In argparse:
```python
parser.add_argument('--tandem_cross_dsdf', action='store_true')
```

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/cross-dsdf-s42" --wandb_group phase6/tandem-cross-dsdf \
  --tandem_cross_dsdf \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "askeladd/cross-dsdf-s73")
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| cross-dsdf | 42 | | | | | | |
| cross-dsdf | 73 | | | | | | |
| **2-seed avg** | — | | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | — | d7l91p0x, j9btfx09 |

**Also report:** Does DSDF feature order match expectation? What is `dist_ratio` mean over tandem samples (sanity check: should be ~0.5 on average)? VRAM delta vs baseline (expect ~0 since no new parameters in the backbone)?

**Debug checkpoint:** If training crashes at model init, check whether `fun_dim` / `X_DIM` is being set correctly for the 2 new channels.

## Baseline

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

Current best: PR #2130 (GSB + PCGrad). W&B: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3).

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Key Metrics (2-seed avg)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 12.6 | 13.2 | **12.9** | 13.05 | **-1.1%** |
| p_oodc | 7.8 | 7.8 | **7.8** | 7.70 | +1.3% |
| **p_tan** | 30.3 | 29.4 | **29.85** | 28.60 | **+4.4%** |
| p_re | 6.7 | 6.5 | **6.6** | 6.55 | +0.8% |

### Val Loss

| Seed | val_in_dist | val_tandem | val_ood_cond | val_ood_re |
|------|-------------|------------|--------------|------------|
| 42 | 0.2544 | 0.9015 | 0.1932 | 0.1976 |
| 73 | 0.2622 | 0.8902 | 0.1914 | 0.1927 |

### VRAM & Training
- Seed 42: 46.1 GB peak (baseline ~46 GB) — **negligible VRAM delta** as expected
- Seed 73: 45.8 GB peak
- Both runs: 150 epochs in 180 minutes (3h timeout), 70-79s/epoch
- Model IDs: `jkm5f1yx` (seed 42), `vgm0gla2` (seed 73)
- W&B group: `phase6/tandem-cross-dsdf`

### Cross-DSDF Feature Sanity Checks
- `dist_ratio` logged to W&B as `train/cross_dsdf_dist_ratio_mean` — values vary across batches (not collapsed), confirming non-degenerate features
- `rel_angle` logged as `train/cross_dsdf_rel_angle_mean_abs` — also shows healthy variation
- Both features zeroed for non-tandem samples via `_tandem_mask_f`
- DSDF channel ordering confirmed: foil-1 at x[:,:,2:6], foil-2 at x[:,:,6:10] (matches `aug_dsdf2_sigma` convention)

### Commands Used

```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/cross-dsdf-s{42,73}" --wandb_group phase6/tandem-cross-dsdf \
  --tandem_cross_dsdf \
  --seed {42,73} \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### What Happened — Negative Result

The cross-DSDF features **did not improve p_tan** and in fact degraded it by +4.4% relative to baseline. Detailed analysis:

1. **p_tan regression (+4.4%)**: The primary target metric worsened from 28.60 → 29.85. Both seeds are consistently above baseline (30.3, 29.4), ruling out seed noise.

2. **p_in slightly improved (-1.1%)**: The in-distribution single-foil metric got marginally better, but this is within noise range.

3. **p_oodc and p_re slightly worse**: Both off-target metrics regressed slightly, suggesting the extra features don't help generalization.

4. **Why it likely failed**: The two cross-DSDF scalars (`dsdf_dist_ratio`, `dsdf_rel_angle`) are **coarse summaries** of the per-node inter-foil geometry. The model already has access to the full 8-channel DSDF field (foils 1+2) and can learn whatever inter-foil relationships it needs directly from those raw channels. The hand-crafted summaries likely add redundant noise rather than genuinely new information. The `min` operation (used for `dsdf_dist_ratio`) is particularly lossy — it discards directional DSDF information that the Transolver attention can already exploit.

5. **Feature design issue**: Using `min(abs(dsdf))` as a proxy for "distance to foil" conflates signed distance information with magnitude, and the angle feature using `atan2` of specific DSDF channels may not align with the most informative gradient direction for the physics.

### Suggested Follow-ups

- **Don't pursue raw cross-DSDF scalars further** — the model has sufficient representational capacity to learn inter-foil relationships from the full DSDF field.
- **If inter-foil geometry features are still desired**, a learned approach (e.g., a small MLP that takes both foils' DSDF vectors and outputs a learned inter-foil embedding) would likely outperform hand-crafted scalars — but this adds parameters and complexity.
- **Focus future tandem improvements on the loss/training side** (PCGrad variants, domain-specific weighting) rather than input feature engineering, since the DSDF channels already provide rich spatial information.
